packages/clerk-js/src/ui/hooks/usePasswordComplexity.ts

```diff
@@ -0,0 +1,126 @@
+import type { PasswordSettingsData } from '@clerk/types';
+import { useCallback, useMemo, useState } from 'react';
+
+import { localizationKeys, useLocalizations } from '../localization';
+
+const testComplexityCases = (
+  password: string,
+  {
+    minLength,
+    maxLength,
+  }: {
+    minLength: number;
+    maxLength: number;
+  },
+) => {
+  return {
+    max_length: password.length < maxLength,
+    min_length: password.length >= minLength,
+    require_numbers: /\d/.test(password),
+    require_lowercase: /[a-z]/.test(password),
+    require_uppercase: /[A-Z]/.test(password),
+    require_special_char: /[!"#$%&'()*+,-./:;<=>?@[\]^_`{|}~]/.test(password),
+  };
+};
+
+type ComplexityErrorMessages = {
+  [key in keyof Partial<Omit<PasswordSettingsData, 'disable_hibp' | 'min_zxcvbn_strength' | 'show_zxcvbn'>>]: string;
+};
+
+export const usePasswordComplexity = (
+  config: Omit<PasswordSettingsData, 'disable_hibp' | 'min_zxcvbn_strength' | 'show_zxcvbn'>,
+  callbacks?: {
+    onValidationFailed?: (validationErrorMessages: ComplexityErrorMessages, errorMessage: string) => void;
+    onValidationSuccess?: () => void;
+  },
+) => {
+  const { min_length, max_length, require_lowercase, require_numbers, require_uppercase, require_special_char } =
```

Comment:
How come these are snake_cased? Aren't we parsing them in UserSettings?

Reply:
It seems that we do. We only to do that for the first level, we do not follow the pattern for all nested objects.

For example
`userSettings.signIn.second_factor` and `userSettings.signUp.allowlist_only`

Reply:
👉https://github.com/clerkinc/javascript/blob/8f190fcb72a9c8c21df66842db9bf26543f98bcb/packages/clerk-js/src/core/resources/UserSettings.ts#L56